evm/src/cpu/kernel/asm/fields/fp12_mul.asm

```diff
@@ -71,96 +82,372 @@
+///    DUP6  %offset_fp6
+///    // stack:                                         inA', g0, G1, G1', inA, inB, out
+///    %load_fp6
+///    // stack:                                           f', g0, G1, G1', inA, inB, out
+///    DUP12
+///    // stack:                                      inA, f', g0, G1, G1', inA, inB, out
+///    %load_fp6
+///    // stack:                                        f, f', g0, G1, G1', inA, inB, out
+///    %clone_mul_fp_fp6
+///    // stack:                                 (g0)f, f, f', g0, G1, G1', inA, inB, out
+///    %clone_mul_fp2_fp6_sh
+///    // stack:                     (G1)sh(f) , (g0)f, f, f', g0, G1, G1', inA, inB, out
+///    %add_fp6
+///    // stack:                     (G1)sh(f) + (g0)f, f, f', g0, G1, G1', inA, inB, out
+///    %clone_mul_fp2_fp6_sh2
+///    // stack:      (G1')sh2(f') , (G1)sh(f) + (g0)f, f, f', g0, G1, G1', inA, inB, out 
+///    %add_fp6
+///    // stack:      (G1')sh2(f') + (G1)sh(f) + (g0)f, f, f', g0, G1, G1', inA, inB, out 
+///    DUP26
+///    // stack: out, (G1')sh2(f') + (G1)sh(f) + (g0)f, f, f', g0, G1, G1', inA, inB, out 
+///    %store_fp6
+///    // stack:                                        f, f', g0, G1, G1', inA, inB, out 
+///    %semiclone_mul_fp2_fp6_sh
+///    // stack:                               (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    %clone_mul_fp2_fp6_sh
+///    // stack:                  (G1)sh(f') , (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    %add_fp6
+///    // stack:                  (G1)sh(f') + (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    %clone_mul_fp_fp6
+///    // stack:         (g0)f' , (G1)sh(f') + (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    %add_fp6
+///    // stack:         (g0)f' + (G1)sh(f') + (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    DUP20  offset_fp6
+///    // stack:   out', (g0)f' + (G1)sh(f') + (G1')sh(f), f', g0, G1, G1', inA, inB, out 
+///    %store_fp6
+///    // stack:                                           f', g0, G1, G1', inA, inB, out 
+///    %pop14
+
+
+/////////////////////////
+///// FP12 SQUARING /////
+/////////////////////////
+
+/// cost: 646
+
+/// fp6 functions:
+///  fn    | num | ops | cost
+///  -------------------------
+///  load  |   2 |  40 |   80
+///  store |   2 |  40 |   80
+///  dup   |   2 |   6 |   12
+///  swap  |   2 |  16 |   32
+///  add   |   1 |  16 |   16
+///  mul   |   1 | 157 |  157
+///  sq    |   2 | 101 |  202
+///  dbl   |   1 |  13 |   13
+///
+/// lone stack operations:
+///  op    | num 
+///  ------------
+///  ADD   |   3
+///  SWAP  |   4
+///  DUP   |   5
+///  PUSH  |   6
+///  POP   |   3
+///  JUMP  |   4
+///
+/// TOTAL: 
+
+/// input:
+///     F = f + f'z
+///
+/// output:
+///     H = h + h'z = FF
+///
+///     h  = ff + sh(f'f')
+///     h' = 2ff'
+///
+/// memory pointers [ind' = ind+6]
+///     {inp: f, inp: f', out: h, out': h'}
+///
+/// f, f' consist of six elements on the stack
```

Comment:
Interestingly, Karatsuba helps with squaring too, even though squaring already involves only three multiplications.

Or rather, the naive implementation of squaring does two squares (`ff` and `f'f'`) and one multiplication (`ff'`).

With Karatsuba, we form
```
h = ff + sh(f'f')
h' = (f+f')(f+f') - ff - f'f'
```

Notice that this is three squares. It may happen that it's not actually worth it to replace a mul with a square due to the overhead of this method (two more additions), but my suspicion is that Karatsuba would be a bit faster.